plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinUastResolveProviderService.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.frontend.api.analyse
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtReferenceExpression
+import org.jetbrains.uast.UElement
+
+interface FirKotlinUastResolveProviderService : BaseKotlinUastResolveProviderService {
+    override fun convertParent(uElement: UElement): UElement? {
+        TODO("Not yet implemented")
+    }
+
+    override fun resolveToDeclaration(ktExpression: KtExpression): PsiElement? {
+        when (ktExpression) {
+            is KtReferenceExpression -> {
+                analyse(ktExpression) {
```

Comment:
Let't use analyseWithCustomToken here. It allows to specify ValidityTokenFactory to use like it is done for light classes. For start, we can use AlwaysAccessibleValidityTokenFactory here

Reply:
Thanks for the pointers. As 3rd commit is more or less refactoring, I'd like to keep the use of `analyse` and migrate that as-is. Instead, added 6th commit that changes the resolution to use `analyseWithCustomToken` with `AlwaysAccessibleValidityTokenFactory`